src/m_argv.c

```diff
@@ -261,6 +261,127 @@ void M_FindResponseFile(void)
     }
 }
 
+void M_AddLooseFiles(void)
+{
+    int i, j, k;
+
+    char **wads;
+    char **lmps;
+    char **dehs;
+    int wadcount = 0;
+    int lmpcount = 0;
+    int dehcount = 0;
+
+    char **newargv;
+    int newargc;
+
+    struct {
+        const char *ext;
+        char ***list;
+        int *count;
+    } looses[] = {
+        {".wad", &wads, &wadcount},
+//      {".lmp", &wads, &wadcount}, // demo lumps are the most common
+        {".deh", &dehs, &dehcount},
+//      {".bex", &dehs, &dehcount},
+        {".hhe", &dehs, &dehcount},
+        {".seh", &dehs, &dehcount},
+        {".lmp", &lmps, &lmpcount},
+    };
+
+    struct {
+        const char *cmdparam;
+        char ***list;
+        int *count;
+    } params[] = {
+        {"-merge"   , &wads, &wadcount},
+        {"-deh"     , &dehs, &dehcount},
+        {"-playdemo", &lmps, &lmpcount},
+    };
+
+    // check for command line arguments
+    if (myargc < 2)
+    {
+        return;
+    }
+
+    // check for regular arguments and response files
+    for (i = 1; i < myargc; i++)
+    {
+        if (myargv[i][0] == '-' || myargv[i][0] == '@')
+        {
+            return;
+        }
+    }
+
+    // we will find at most myargc files of a kind
+    for (j = 0; j < arrlen(params); j++)
+    {
+        *(params[j].list) = malloc(myargc * sizeof(char *));
+    }
+
+    // identify loose files on the command line by their extensions
```

Comment:
Slightly different idea here but tell me what you think of it:

The goal here is to basically arrange the list by file type (using extension to identify file type). So one alternative approach you could use would be to treat it as a sorting problem: sort the myargv[1:] list, where the sort key is produced by a function that takes a filename and produces a file type. For example:
```c
enum {
  FILETYPE_UNKNOWN,
  FILETYPE_IWAD,
  FILETYPE_PWAD,
  FILETYPE_DEH,
}
```
If you did it this way, then after sorting you get files grouped by order, you just need to insert the `-file` etc. where appropriate, and it avoids having to build up separate lists.

Up to you if you want to just stick with the current approach but I figured I'd throw it out there.

Reply:
Neat idea turning this into a sorting problem! Still, the crucial point is to insert the appropriate parameters into the sorted list. I'll have to think about it...

Reply:
Do we expect `qsort()` to be available or should I add a check to `configure.ac`?

Reply:
Wait, there's `SDL_qsort()` that I can rely on.